leancodepl/patrol
# Preliminary support for running on physical iOS devices with `ios-deploy`

Aimed at resolving #829

Also addresses #853 

## Why is this needed?

Since iOS 14, Flutter apps can't be run in debug mode "normally", but through a debugger. This can be done by running directly from Xcode or by using the unofficial `ios-deploy` tool (which is also used by `flutter_tools` under the hood).

## How to run Patrol test on a real iPhone?

> These steps will all be done by `patrol test`. I'm listing them here for _expanding and preserving our knowledge_'s sake.

1. Build the apps: _app under test_ (target Runner) and _test runner app_ (target RunnerUITests):

   ```console
   xcodebuild build-for-testing \
     -workspace Runner.xcworkspace \
     -scheme Runner \
     -xcconfig Flutter/Debug.xcconfig \
     -configuration Debug \
     -sdk iphoneos -destination "generic/platform=iOS" \
     -quiet \
     -derivedDataPath ../build/ios_integ \
     OTHER_SWIFT_FLAGS='$(inherited) -D PATROL_ENABLED'
   ```

2. Install the app under test (run from `ios` directory):

   ```console
   ios-deploy \
     --id 00008101-001611D026A0001E \
     --bundle ../build/ios_integ/Build/Products/Debug-iphoneos/Runner.app \
     --app_deltas ../build/ios_integ/Runner-delta \
     --debug --nostart --no-wifi \
     --args "--enable-dart-profiling --enable-checked-mode --verify-entry-points"
   ```

   This opens LLDB console and waits for commands.

3. Run the app under test

   Type `run` in the LLDB console to run the app under test. The app under test
   starts up and every 1 second (for 60 seconds) tries to contact the automation
   server (which lives in the test runner app - we'll start it in the next step).

   Once the automation server gets up and responds, tests start.

4. Start UI tests through `xcodebuild`

   ```console
   xcodebuild test-without-building \
     -xctestrun "../build/ios_integ/Build/Products/Runner_iphoneos16.2-arm64.xctestrun" \
     -destination "platform=iOS,name=Barteks-iPhone"	`# replace with your device name` \
     -resultBundlePath TestResults
   ```
    This also installs the test runner app.

## Problems

### `ios-deploy` sometimes gets stuck

Rarely, after executing:

```console
ios-deploy \
  --id 00008101-001611D026A0001E \
  --bundle ../build/ios_integ/Build/Products/Debug-iphoneos/RunnerUITests-Runner.app \
  --app_deltas ../build/ios_integ/Runner-delta \
  --debug --nostart --no-wifi \
  --args "--enable-dart-profiling --enable-checked-mode --verify-entry-points"
```

it gets stuck somewhere at the end of "copying resources" phase:

```
...
[ 49%] Copying /Users/bartek/dev/leancode/patrol/packages/patrol/example/build/ios_integ/Build/Products/Debug-iphoneos/Runner.app/embedded.mobileprovision to device
[ 49%] Copying /Users/bartek/dev/leancode/patrol/packages/patrol/example/build/ios_integ/Build/Products/Debug-iphoneos/Runner.app/Info.plist to device
[ 49%] Copying /Users/bartek/dev/leancode/patrol/packages/patrol/example/build/ios_integ/Build/Products/Debug-iphoneos/Runner.app/PkgInfo to device
```

Clearing `ios-deploy` delta directories and restarting my iPhone reliably solves this issue. Also, try unplugging the cable and plugging again.

### `ios-deploy` sometimes segfaults

And I've no idea why. Usually uninstalling the apps, restarting the iPhone/iPad helps.

I think we should research replacing [ios-deploy](https://github.com/ios-control/ios-deploy) with [go-ios](https://github.com/danielpaulus/go-ios).

`ios-deploy` is basically [a 3000+ lines Objective-C file](https://github.com/ios-control/ios-deploy/blob/master/src/ios-deploy/ios-deploy.m).

`go-ios` is a nice, somewhat better tested, pure Go project. 

## Additional context

Might be useful:
- https://docs.flutter.dev/development/add-to-app/ios/project-setup#local-network-privacy-permissions

See also:

- https://github.com/flutter/flutter/issues/60657#issuecomment-688478590
